### ISSUE TITLE
When I execute socket.add() or socket.write(), an error is reported: SocketException: OS Error: Software caused connection abort, errno = 103

### ISSUE THREAD

**Issue by jing201151603** [flutter/flutter#84839]
This is my code:

```dart
try {
      LogUtils.log(
          tag, "isLogin=$isLogin, socket=$socket,msgCode=$msgCode");

      socket.add(msg);
    } catch (e) {
      print("msgCode=${msgCode},e=${e.toString()}");
    }
```
When my program is executed here, my network status is good, but when this method socket.add() is executed, my network is suddenly disconnected, and the console will report an error: [ERROR:flutter/runtime/dart_isolate .cc(1137)] Unhandled exception:
E/flutter (25901): SocketException: OS Error: Software caused connection abort, errno = 103.
And the application crashed, try catch did not take effect, and the print() function did not execute. In theory, I will not crash after catch, so I think this is a bug of flutter sdk, you have a better solution Way?

**Comment by jing201151603** [comment]
The Socket class is in the package'dart:io'.

**Comment by TahaTesser** [comment]
Hi @jing201151603 
Can you please provide your `flutter doctor -v`, a complete reproducible minimal code sample, and complete `flutter run --verbose` logs running the minimal code sample?
Thank you

**Comment by jing201151603** [comment]
> Hi @jing201151603
> Can you please provide your `flutter doctor -v`, a complete reproducible minimal code sample, and complete `flutter run --verbose` logs running the minimal code sample?
> Thank you

Hi, this is my code:

import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

class Test {
  Socket socket;

  Timer _timerKeepLive;

  //  心跳发送间隔
  int _keepAliveInterval = 2 * 1000;

  //心跳超时间隔
  int _keepAliveOutInterval = 5 * 1000;

  //最后一次收到心跳的时间
  int _lastKeepAliveRecvTime;

  Future connect(severIp, port, String name, String pwd) async {
    try {
      socket =
          await Socket.connect(severIp, port, timeout: Duration(seconds: 2));

      await _login(name, pwd);
      streamListen();
    } catch (e) {
      print(e);
      return;
    }
  }

  Future _login(String name, String pwd) {
    var req = LibPkg.LoginMessage(name, pwd, 0x000a6a);
    sendLoginRequest(LibPkg.NET_CMD_LOGIN, req);
  }

  void sendLoginRequest(int msgCode, LibPkg.LoginMessage pb) {
    var username =
        unicode2gbk(utf82unicode(utf8.encode(pb.userName?.toUpperCase())));
    m_userName = utf8.decode(Uint8List.fromList(username));
    var userNameLen = ByteData(4);
    userNameLen.setUint32(0, username.length);

    String strPasswd = sha1.convert(utf8.encode(pb.passWord)).toString();
    var passwd = utf8.encode(strPasswd.toUpperCase());
    var passwdLen = ByteData(4);
    passwdLen.setUint32(0, passwd.length);

    var version = ByteData(4);
    version.setInt32(0, pb.version);

    Uint8List pbBody;
    pbBody = Uint8List.fromList(userNameLen.buffer.asUint8List() +
        username +
        passwdLen.buffer.asUint8List() +
        passwd +
        version.buffer.asUint8List());

    LibPkg.FlutterPackage fpk = LibPkg.FlutterPackage();
    fpk.head.cmd_i16 = msgCode;
    fpk.body = pbBody;
    sendMsg(fpk);
  }

  //发消息
  void sendMsg(LibPkg.FlutterPackage fpk) {
    //包头+包体组合成一个完整的数据包
    var msg = fpk.compressEncode();
    var msgCode = fpk.head.cmd_i16;

    if (socket == null) {
      return;
    }

    //给服务器发消息
    try {
      socket?.add(msg);
    } catch (e) {
      doneHandler();
    }
  }

  void streamListen() {
    socket.listen(decodeHandle,
        onError: errorHandler, onDone: doneHandler, cancelOnError: false);
  }

  void decodeHandle(Uint8List newData) {
    onKeepAlive(true);
  }

  void doneHandler() {
    if (socket == null) return;
    socket.close();
    socket.destroy();
    socket = null;
  }

  void errorHandler(error, StackTrace trace) {
    doneHandler();
  }

  void onKeepAlive(bool bRecv) {
    if (bRecv) {
      // 收到心跳信息，更新时间
      _lastKeepAliveRecvTime = DateTime.now().millisecondsSinceEpoch;
    } else {
      // 发送心跳信息
      if (_timerKeepLive != null) _timerKeepLive.cancel();
      _timerKeepLive =
          Timer.periodic(Duration(milliseconds: _keepAliveInterval), (_timer) {
        int nowt = DateTime.now().millisecondsSinceEpoch;
        //time out
        if (nowt - _lastKeepAliveRecvTime > _keepAliveOutInterval) {
          _timer.cancel();
          doneHandler();
        } else {
          try {
            sendKeepLive(0);
          } catch (e) {
            doneHandler();
          }
        }
      });
    }
  }

  void sendKeepLive(int msgCode) {
    var pbBody = ByteData(4);
    var now = DateTime.now();
    pbBody.setInt32(0, now.millisecondsSinceEpoch);
    LibPkg.FlutterPackage fpkg = LibPkg.FlutterPackage();
    fpkg.head.cmd_i16 = msgCode;
    fpkg.body = pbBody.buffer.asUint8List();
    var msg = fpkg.encode();
  try {
      //LogUtils.log(
        //  tag, "isLogin=$isLogin, socket=$socket,msgCode=$msgCode");

      socket.add(msg);
    } catch (e) {
      //print("msgCode=${msgCode},e=${e.toString()}");
      print(e);
    }
  }
}



My flutter version is Flutter (Channel stable, 2.2.2, on macOS 11.4 20F71 darwin-x64, locale zh-Hans-CN).
Not sure if this is what you want, if you need anything, you can ask me.
Thank you

**Comment by jing201151603** [comment]
> Hi @jing201151603
> Can you please provide your `flutter doctor -v`, a complete reproducible minimal code sample, and complete `flutter run --verbose` logs running the minimal code sample?
> Thank you

Hi, this is My code:

import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

class Test {
  Socket socket;

  Timer _timerKeepLive;

  int _keepAliveInterval = 2 * 1000;

  int _keepAliveOutInterval = 5 * 1000;

  int _lastKeepAliveRecvTime;

  Future connect(severIp, port, String name, String pwd) async {
    try {
      socket =
          await Socket.connect(severIp, port, timeout: Duration(seconds: 2));

      await _login(name, pwd);
      streamListen();
    } catch (e) {
      print(e);
      return;
    }
  }

  Future _login(String name, String pwd) {
    sendLoginRequest(name, pwd);
  }

  void sendLoginRequest(String name, String code) {
    var fpk = "$name-$code";
    sendMsg(fpk);
  }

  void sendMsg(dynamic fpk) {
    var msg = fpk.compressEncode();

    if (socket == null) {
      return;
    }

    try {
      //LogUtils.log(
      //tag, "isLogin=$isLogin, socket=$socket,msgCode=$msgCode");

      socket.add(msg);
    } catch (e) {
      doneHandler();
      print("e=${e.toString()}");
    }
  }

  void streamListen() {
    socket.listen(decodeHandle,
        onError: errorHandler, onDone: doneHandler, cancelOnError: false);
  }

  void decodeHandle(Uint8List newData) {
    onKeepAlive(true);
  }

  void doneHandler() {
    if (socket == null) return;
    socket.close();
    socket.destroy();
    socket = null;
  }

  void errorHandler(error, StackTrace trace) {
    doneHandler();
  }

  void onKeepAlive(bool bRecv) {
    if (bRecv) {
      _lastKeepAliveRecvTime = DateTime.now().millisecondsSinceEpoch;
    } else {
      if (_timerKeepLive != null) _timerKeepLive.cancel();
      _timerKeepLive =
          Timer.periodic(Duration(milliseconds: _keepAliveInterval), (_timer) {
        int nowt = DateTime.now().millisecondsSinceEpoch;
        //time out
        if (nowt - _lastKeepAliveRecvTime > _keepAliveOutInterval) {
          _timer.cancel();
          doneHandler();
        } else {
          try {
            sendKeepLive(0);
          } catch (e) {
            doneHandler();
          }
        }
      });
    }
  }

  void sendKeepLive(int msgCode) {
    var pbBody = ByteData(4);
    var now = DateTime.now();
    pbBody.setInt32(0, now.millisecondsSinceEpoch);
    var fpkg = LibPkg.FlutterPackage();
    fpkg.head.cmd_i16 = msgCode;
    fpkg.body = pbBody.buffer.asUint8List();
    var msg = fpkg.encode();
    try {
      //LogUtils.log(
      //tag, "isLogin=$isLogin, socket=$socket,msgCode=$msgCode");

      socket.add(msg);
    } catch (e) {
      doneHandler();
      print("msgCode=$msgCode,e=${e.toString()}");
    }
  }
}


My flutter -v is Flutter (Channel stable, 2.2.2, on macOS 11.4 20F71 darwin-x64, locale zh-Hans-CN).
Not sure if this is what you want, if you need anything, you can ask me.
Thank you.

**Comment by TahaTesser** [comment]
Hi @jing201151603 
Looks like your code contains unrelated code, please remove additional code and provide a complete runnable sample

**Comment by jing201151603** [comment]
> Looks like your code contains unrelated code, please remove additional code and provide a complete runnable sample

Sorry, our network communication is data encrypted, but this should have nothing to do with data communication, but socket.add, the network is unblocked at this moment, but it is possible that the user's network status is suddenly abnormal, at this time socket.add The data sent did not seem to be sent, which caused a crash, and try catch did not take effect, still causing the entire APP to crash. From the console log, it can be located that this line of code was caused: socket.add(); the following is the crash 'S log:

E/flutter (12390): [ERROR:flutter/runtime/dart_isolate.cc(1137)] Unhandled exception:
E/flutter (12390): SocketException: Write failed (OS Error: Software caused connection abort, errno = 103), address = 175.25.17.144, port = 49376
E/flutter (12390): #0      _NativeSocket.write (dart:io-patch/socket_patch.dart:1114:34)
E/flutter (12390): #1      _RawSocket.write (dart:io-patch/socket_patch.dart:1759:15)
E/flutter (12390): #2      _Socket._write (dart:io-patch/socket_patch.dart:2192:18)
E/flutter (12390): #3      _SocketStreamConsumer.write (dart:io-patch/socket_patch.dart:1940:26)
E/flutter (12390): #4      _SocketStreamConsumer.addStream.<anonymous closure> (dart:io-patch/socket_patch.dart:1914:11)
E/flutter (12390): #5      _RootZone.runUnaryGuarded (dart:async/zone.dart:1546:10)
E/flutter (12390): #6      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
E/flutter (12390): #7      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
E/flutter (12390): #8      _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
E/flutter (12390): #9      _StreamController._add (dart:async/stream_controller.dart:607:7)
E/flutter (12390): #10     _StreamController.add (dart:async/stream_controller.dart:554:5)
E/flutter (12390): #11     _StreamSinkImpl.add (dart:io/io_sink.dart:136:17)
E/flutter (12390): #12     _Socket.add (dart:io-patch/socket_patch.dart:2039:38)
E/flutter (12390): #13     Test.sendKeepLive (package:tl_net/src/test.dart:1066:15)
E/flutter (12390): #14     Test.onKeepAlive.<anonymous closure> (package:tl_net/src/test.dart:131:17)
E/flutter (12390): #15     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:395:19)
E/flutter (12390): #16     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:426:5)
E/flutter (12390): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)

**Comment by jing201151603** [comment]
> Hi @jing201151603
> Looks like your code contains unrelated code, please remove additional code and provide a complete runnable sample

Sorry, our network communication is data encrypted, but this should have nothing to do with data communication, but socket.add, the network is unblocked at this moment, but it is possible that the user's network status is suddenly abnormal, at this time socket.add The data sent did not seem to be sent, which caused a crash, and try catch did not take effect, still causing the entire APP to crash. From the console log, it can be located that this line of code was caused: socket.add(); the following is the crash 'S log:

E/flutter (12390): [ERROR:flutter/runtime/dart_isolate.cc(1137)] Unhandled exception:
E/flutter (12390): SocketException: Write failed (OS Error: Software caused connection abort, errno = 103), address = 127.0.0.1, port = 46739
E/flutter (12390): #0 _NativeSocket.write (dart:io-patch/socket_patch.dart:1114:34)
E/flutter (12390): #1 _RawSocket.write (dart:io-patch/socket_patch.dart:1759:15)
E/flutter (12390): #2 _Socket._write (dart:io-patch/socket_patch.dart:2192:18)
E/flutter (12390): #3 _SocketStreamConsumer.write (dart:io-patch/socket_patch.dart:1940:26)
E/flutter (12390): #4 _SocketStreamConsumer.addStream. (dart:io-patch/socket_patch.dart:1914:11)
E/flutter (12390): #5 _RootZone.runUnaryGuarded (dart:async/zone.dart:1546:10)
E/flutter (12390): #6 _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
E/flutter (12390): #7 _BufferingStreamSubscription._add (dart:async/stream_impl.dart:271:7)
E/flutter (12390): #8 _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:733:19)
E/flutter (12390): #9 _StreamController._add (dart:async/stream_controller.dart:607:7)
E/flutter (12390): #10 _StreamController.add (dart:async/stream_controller.dart:554:5)
E/flutter (12390): #11 _StreamSinkImpl.add (dart:io/io_sink.dart:136:17)
E/flutter (12390): #12 _Socket.add (dart:io-patch/socket_patch.dart:2039:38)
E/flutter (12390): #13 Test.sendKeepLive (package:tl_net/src/test.dart:1066:15)
E/flutter (12390): #14 Test.onKeepAlive. (package:tl_net/src/test.dart:131:17)
E/flutter (12390): #15 _Timer._runTimers (dart:isolate-patch/timer_impl.dart:395:19)
E/flutter (12390): #16 _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:426:5)
E/flutter (12390): #17 _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)

**Comment by TahaTesser** [comment]
Hi @jing201151603 
if it is not your network, please provide a reproducible code, and possibly add  any public test link 


> Looks like your code contains unrelated code, please remove additional code and provide a complete runnable sample

**Comment by TahaTesser** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.
